### ISSUE TITLE
Fix `property-(dis)allowed-list` false negatives for custom properties

### ISSUE THREAD

**Issue by fbasmaison-lucca** [stylelint/stylelint#8187]
### What is the problem you're trying to solve?

When using the `property-disallowed-list`, I was expecting custom properties to be disallowed too.

[Those lines](https://github.com/stylelint/stylelint/blob/b2c99cef764643f3bd9539b34cdec58af882db88/lib/rules/property-disallowed-list/index.mjs#L39-L41) seem to tell otherwise.

Was there a reason to not include custom properties in the mix?

### What solution would you like to see?

I might misunderstand basic naming convention, so ignore custom properties may be normal.

If it is not the case, I would expect this rule to:

- Take all properties into consideration (standardised and custom).
- Have an option to ignore standardised or custom properties.

Otherwise, another rule can be specific to custom properties.

I probably can code either, so this Issue is merely to know which path to follow.

**Comment by ybiquitous** [comment]
Thanks for opening this discussion. I think it makes sense to exclude custom properties in [`property-disallowed-list`](https://stylelint.io/user-guide/rules/property-disallowed-list) and use [`custom-property-pattern`](https://stylelint.io/user-guide/rules/custom-property-pattern) instead. In short, I don't think it's necessary to change the current behavior.

However, I agree with updating the `property-disallowed-list` document to introduce `custom-property-pattern`. For example:

```
This rule doesn't work for custom properties. Use the `custom-property-pattern` rule instead.
```

We may also be able to say the same for [`property-allowed-list`](https://stylelint.io/user-guide/rules/property-allowed-list).

**Comment by fbasmaison-lucca** [comment]
I wouldn’t mind an explanation about why you think it makes sense to exclude custom properties, by the way. :)

**Comment by Mouvedia** [comment]
> […and added a check for --custom-properties, figuring […] we should ignore those as well.](https://github.com/stylelint/stylelint/pull/520#issuecomment-158651602)

see c9d1967

**Comment by jeddy3** [comment]
Oh wow, pre `3.x`. That's going back a long time...

> Having to juggle with regexp to disallow specific custom properties does not seem ideal to me. Our use case is for deprecated variables in a Design System.

It's a sound use case.

([`custom-property-pattern`](https://stylelint.io/user-guide/rules/custom-property-pattern) is complementary and ensures properties follow a pattern, such as kebab-case or with a prefix.)

The behaviour of the `property-disallowed-list` and `property-allowed-list` rules is inconsistent with our other `*-property-*` rules, given:

```css
a {
  --var: 10px;
}
```

With this config:

```json
{
  "rules": {
    "declaration-property-value-disallowed-list": {
      "/^--/": ["10px"]
    },
    "declaration-property-unit-disallowed-list": {
      "/^--/": ["px"]
    },
    "rule-selector-property-disallowed-list": {
      "a": "/^--/"
    }
  }
}
```

We get:

```console
2:2-7 error Unexpected property "--var" for selector "a" (rule-selector-property-disallowed-list)
2:9-13 error Unexpected unit "px" for property "--var" (declaration-property-unit-disallowed-list)
2:9-13 error Unexpected value "10px" for property "--var" (declaration-property-value-disallowed-list)
```

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdugLgWjwDckAneFARgAYAHADwG4sBfEAGhADMBLAGxgBySALZxEMOiJr8AdGADO89uAiYeAcwQgMmFCnQgSAV37yD5HXr0HYYXqSQ5uqvDRIQaMEjgCeRJLxGMHhQ3PIBvBAA7jBQeLxhOOZoWFZWBgD0AHoEGckA2gbU9AYAuqlWzGwV1iC29iSOzpiu7p7efkaY3Pih4byRMXEJ8kkIKbpp+iDZuQUGJSDlk3pVNdPG-HjyMPxgOBAkrR5eviFhEdGx8YnJllMGSMmZOXh5IOvMFV+YrByQam46gAYodhI4tAAreSqZSwGhKRD3AyjHz8BKYMbkAz2HAwUYGaqTFG+dHcTF4AEabY4JCYKCkKDPEC4-FjFggZhAA).

I recently relied on this behaviour to demonstrate how users can limit how they define their custom properties (https://github.com/stylelint/stylelint/issues/7974#issuecomment-2328613984).

Rather than add a `custom-property-disallowed-list` rule to support @fbasmaison-lucca use case, shall we bring `property-disallowed-list` and `property-allowed-list` in line with the other `property` rules?

**Comment by ybiquitous** [comment]
> Rather than add a `custom-property-disallowed-list` rule to support [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/144820795?s=64&amp;u=89d485f7870701509395a02301780dfadba0538c&amp;v=4">@fbasmaison-lucca](https://github.com/fbasmaison-lucca) use case, shall we bring `property-disallowed-list` and `property-allowed-list` in line with the other `property` rules?

That makes sense 👍🏼

**Comment by jeddy3** [comment]
@fbasmaison-lucca I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.
